docs/design/provisioning.md

```diff
@@ -0,0 +1,322 @@
+* Allow admins to configure many recyclers of persistent volumes
+* Decouple provisioning code from volume plugin code
+
+
+### PersistentStorageConfig
+
+Administrators map binders, provisioners, and recyclers to specific types of persistent claims and volumes by creating a PersistentStorageConfig file.  The config file is set as a CLI flag.
+
+Binders contains two selectors against which they match claims for binding.  One selector matches claims and the other selector matches volumes.  Both must be true in order to bind, in addition to other matching semantics (e.g, capacity). Claims can remain unbound indefinitely.
+
+Provisioners create resources on demand for claims matching its selector. PersistentVolumeClaims with matching labels will be dynamically provisioned by a plugin named in the config.  Volume attributes are copied from a template.  A specific security context can be assigned to the provisioner.  Volumes created by the provisioner automatically bind to the claim for which it was provisioned.
+
+Recyclers have a selector against which they match persistent volumes for recycling.  A specific security context can be assigned to the recycler. Volumes are in Released state after their bound claims are deleted.  Recyclers may delete volumes or reset phase back to Available, according to each plugin.  Volumes without a matching recycler may be stay Released indefinitely.
+
+
+```
+
+// PersistentStorageConfig is the config defining the management of persistent volumes.
+type PersistentStorageConfig struct {
+	Binders      []BinderConfig
+	Recyclers    []RecyclerConfig
+	Provisioners []ProvisionerConfig
+}
+
+// BinderConfig maps persistent volume claims to volumes when binding claims to available resources
+type BinderConfig struct {
+	// PersistentVolumeClaimSelector is a selector which must be true for the binder to match the claim
+	PersistentVolumeClaimSelector Selector
+	// PersistentVolumeSelect is a selector which must be true for the binder to match the volume
+	PersistentVolumeSelector Selector
+}
+
+type RecyclerConfig struct {
+	// PersistentVolumeSelector is a selector which must be true for the recycler to match the persistent volume
+	PersistentVolumeSelector Selector
+	// RecyclerPluginName is the name of the plugin that can recycle the matched persistent volume.
+	RecyclerPluginName string
+	// RecyclerPodTemplate is a pod template from which new recycling pods are created.
+	// This pod template takes precendence of Driver/Args if both are set
+	RecyclerPodTemplate *api.Pod
+	// SecurityContextName is the name of the security context to use when performing recycling operations
+	SecurityContextName string
+	// Driver is the name of the plugin to use for this Recycle or the name of the driver located beneath
+	//   /usr/libexec/kubernetes/controllermanager-plugins/provisioning/exec
+	// A Recycler must specify a driver/plugin name.
+	Driver string
```

Comment:
We're currently using the `RecyclerPodTemplate` right?  If so, does this have to be in this particular pull?  It seems like it would add something to argue about without providing a lot of power.